src/Controller/WebhookController.php

```diff
@@ -0,0 +1,109 @@
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\EqualsFilter;
+use Shopware\Core\Framework\Validation\DataBag\RequestDataBag;
+use Shopware\Core\System\SalesChannel\SalesChannelContext;
+use Shopware\Storefront\Controller\StorefrontController;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Routing\Annotation\Route;
+use Exception;
+
+/**
+ * @Route(defaults={"_routeScope"={"storefront"}, "csrf_protected"=false})
+ */
+class WebhookController extends StorefrontController
+{
+    /**
+     * @var ClientApiService
+     */
+    private ClientApiService $clientApiService;
+
+    /**
+     * @var OrderTransactionStateHandler
+     */
+    private OrderTransactionStateHandler $orderTransactionStateHandler;
+
+    /**
+     * @var EntityRepositoryInterface
+     */
+    private EntityRepositoryInterface $orderTransactionRepository;
+
+    public function __construct(
+        ClientApiService $clientApiService,
+        OrderTransactionStateHandler $orderTransactionStateHandler,
+        EntityRepositoryInterface $orderTransactionRepository
+    ) {
+        $this->clientApiService = $clientApiService;
+        $this->orderTransactionStateHandler = $orderTransactionStateHandler;
+        $this->orderTransactionRepository = $orderTransactionRepository;
+    }
+
+    /**
+     * @param $transactionId
+     * @param SalesChannelContext $salesChannelContext
+     * @return mixed
+     *
+     * @throws Exception
+     */
+    private function getOrderTransaction($transactionId, SalesChannelContext $salesChannelContext)
+    {
+        $criteria = new Criteria();
+        $criteria->addFilter(new EqualsFilter('customFields.' . CoinGatePaymentHandler::CUSTOM_FIELD_MAPPING_NAME, $transactionId));
+
+        $orderTransactionCollection = $this->orderTransactionRepository->search($criteria, $salesChannelContext->getContext());
+        if ($orderTransactionCollection->count() === 0) {
+            throw new Exception('Order transaction not found for CoinGate order #' . $transactionId);
+        }
+
+        return $orderTransactionCollection->first();
+    }
+
+    /**
+     * @Route("/coingate-payment/webhook", name="storefront.coingate.webhook", methods={"POST","GET"})
+     *
+     * @param RequestDataBag $dataBag
+     * @param Request $request
+     * @param SalesChannelContext $salesChannelContext
+     * @return Response
+     *
+     * @throws Exception
+     */
+    public function index(RequestDataBag $dataBag, Request $request, SalesChannelContext $salesChannelContext): Response
+    {
+        try {
+            $transactionId = $request->get('id', null);
+
+            $client = $this->clientApiService->get($salesChannelContext);
+
+            $order = $client->order->get($transactionId);
+
+            $orderTransaction = $this->getOrderTransaction($transactionId, $salesChannelContext);
```

Comment:
are we not running into the same problem as with whmcs, we should check if payment method is still coingate to avoid overriding another payment